### PR TITLE
Fix #4422, use grad instead of grads[0] for cudnn_batch_norm derivative.

### LINKED ISSUE #4422: the derivative for 'cudnn_batch_norm_backward save_mean' is not implemented
```
loss_gp = ..
loss_gp.backward()  # error
```

Having this error when BatchNorm + wgan_gp (second order) is used.
It's installed from master fec3d4.

### PR DESCRIPTION
**Pull request by ezyang**
The fix proper is very simple: use 'grad' instead of 'grads[0]' in
cudnn_batch_norm's backward definition.

However, the problem was annoyingly subtle and I nearly went off
and implemented a big pile of infrastructure to "tell" the codegen
how to distinguish between differentiable and non-differentiable
outputs before realizing that there must be a way we do this
legitimately, because it works for THNN.  I documented this in
derivatives.yaml, and also added tests for the problem in
load_derivatives.py to catch the various ways you could "get it
wrong".  Hope this helps someone else.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

CC @elbaro

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -3233,6 +3233,13 @@ def func(root):
         gradcheck(func, [v])
         gradgradcheck(func, [v])
 
+    def test_batchnorm_double_backward(self):
+        bn = nn.BatchNorm1d(5).cuda()
```
Comment (colesbury) [inline test/test_nn.py]: The CUDA calls should be guarded by a check or a `@unittest.skipIf`.

Are we doing gradgradchecks of BatchNorm? For both THNN and cuDNN?

Comment (ezyang) [reply test/test_nn.py]: Good catch, I forgot to reenable the tests I disabled >.>

File: tools/autograd/load_derivatives.py [hunk 2]
```diff
@@ -129,6 +129,54 @@ def set_up_derivatives(defn, declaration):
                 args.append(name)
             derivatives.append(create_derivative(declaration, formula, output_indices, args))
 
+        # Test to see if the use of 'grads' makes sense.
+        used_grad = 0
```
Comment (apaszke) [inline tools/autograd/load_derivatives.py]: This is a large chunk of the outer function, can we move it out somewhere else with an appropriate name, so we don't make this one much less readable?
